Ng 49 cleanup optimization

cfboy opened the pull request.

cfboy committed:
Change toast when the program is stopped.

cfboy committed:
Revert. Change toast when the program is stopped. Add comments and documentation.

cfboy committed:
Indent code.

cfboy committed:
Remove unnecessary variable.

cfboy committed:
Merge branches 'NG_49_cleanup_optimization' and 'master' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

 Conflicts:
	utils.py

cfboy committed:
Optimize the position of the seven segment display for easy movement on X or Y axis.

cfboy committed:
Merge branches 'NG_49_cleanup_optimization' and 'master' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

cfboy committed:
add dp on positions

cfboy committed:
Optimize position of I/Os for easy movement on X or Y axis.

cfboy committed:
Indent lines.

cfboy committed:
Change theme color and other improvements.

cfboy committed:
Merge branches 'NG_44_microprocessor_simulator_tests' and 'NG_49_cleanup_optimization' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

 Conflicts:
	GUI/window.py
	microprocessor_simulator.py

ChristianPerez34 commented on GUI/semref.kv:
```diff
@@ -28,64 +28,64 @@
         
         Color:
-            rgb: 1,.8,0
+            rgb: self.border
```
`rgb: self.border_color`

ChristianPerez34 commented on GUI/semref.kv:
```diff
@@ -95,128 +95,133 @@
 <SevenSegmentDisplay>
     canvas.before:
+        Color:
+            rgb: self.border
```
`rgb: self.border_color`

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -45,6 +45,7 @@
 editor_saved = False
 cleared = True
 is_obj = False
+theme_manager = ThemeManager()
```
Can be accessed from from application. Should not be a global variable

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -969,13 +969,18 @@ def get_data(self, address, instruction):
 
 
 class TrafficLights(Widget):
+    global theme_manager
```
Can be accessed from app. Not needed

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -969,13 +969,18 @@ def get_data(self, address, instruction):
 
+    box_pos_x = dp(850)
+    box_pos_y = dp(120)
+    border = ListProperty(theme_manager.primary_color)
```
should be inside `__init__`

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -1064,6 +1069,7 @@ def change_color(self, binary):
 
 
 class SevenSegmentDisplay(Widget):
+    global theme_manager
```
Can be accessed from app. Not needed

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -1079,6 +1085,12 @@ class SevenSegmentDisplay(Widget):
     rightE = ListProperty([.41, .41, .41])
     rightF = ListProperty([.41, .41, .41])
     rightG = ListProperty([.41, .41, .41])
+    border = ListProperty(theme_manager.primary_color)
```
theme manager can be accessed from app. You might need to move this inside of the `__init__` method

`border_color = ListProperty(theme_manager.primary_color)`

ChristianPerez34 commented on utils.py:
```diff
@@ -12,6 +12,12 @@ def convert_to_hex(num, bits):
 
 def convert_to_binary(num, bits):
+    """
+    Converts int number to hexadecimal
```
Converts number to its binary representation

ChristianPerez34 commented on utils.py:
```diff
@@ -12,6 +12,12 @@ def convert_to_hex(num, bits):
+    Converts int number to hexadecimal
+    :param num:
+    :param bits:
+    :return: str:
```
:param num: int
      :param bits: str
      :return: str

ChristianPerez34 commented on utils.py:
```diff
@@ -161,34 +173,35 @@ def update_indicators(instance, is_file_loaded):
     'port': 2
 }
 
+# This ports are reserved by a I/O devices.
```
Reserved ports for I/O devices.

ChristianPerez34 reviewed (changes requested):
Pull from master. Will comment further recommendations once branch is up-to-date

cfboy committed:
Merge branch 'master' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

cfboy committed:
Bug fixes and complete requested changes.

ChristianPerez34 commented on GUI/window.py:
```diff
@@ -12,7 +15,7 @@
 from kivy.graphics.vertex_instructions import Line, Rectangle
 from kivy.metrics import MetricsBase, dp, sp
-from kivy.properties import ListProperty
+from kivy.properties import ListProperty, get_color_from_hex
```
from kivy.utils import get_color_from_hex

Must also run the pycharm optimize imports feature

ChristianPerez34 commented on utils.py:
```diff
@@ -65,17 +70,22 @@ def update_reserved_ports(device, port_to_remove, port_to_add, reserve_block=Fal
+    """
+    This method manage Loaded/NotLoaded File indicators using a condition.
+    :param instance:
+    :param is_ram_loaded:
```
:param instance: obj
:param is_ram_loaded: bool

ChristianPerez34 reviewed (changes requested):
Doesn't conform to upper case naming style
```Python
file_path = ''
can_write = False
loaded_file = False
run_editor = True
editor_saved = False
cleared = True
is_obj = False
```
`HexKeyboard`
simplify if statements

```Python
            for i in range(16):
                if i < 4:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(357), dp(87), dp(35)), width=dp(0.8))
                elif i >= 4 and i < 8:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(322), dp(87), dp(35)), width=dp(0.8))
                elif i >= 8 and i < 12:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(287), dp(87), dp(35)), width=dp(0.8))
                else:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(252), dp(87), dp(35)), width=dp(0.8))
```

`def write_ram(self)`, `def check_loader(self, dt)`, `def open_keyboard(self, instance)`, `def update_io(self, dt)`, `def run_micro_instructions(self, instance)`, `def run_micro_instructions_step(self, instance)`, `def assembler(self)`, `def run_micro_sim(self, file)` missing docstring

`def run_micro_instructions` edit for loop to a while loop since m is not being used
```Python
                for m in range(2):
                    if self.first_inst:
                        self.run_window.inst_table.data_list.clear()
                        self.run_window.inst_table.get_data(self.micro_sim.index,
                                                            self.micro_sim.disassembled_instruction())
                        self.run_window.inst_table.get_data(self.micro_sim.index,
                                                            self.micro_sim.disassembled_instruction())
                        self.first_inst = False
                    else:
                        self.micro_sim.prev_index = -1
                        self.run_window.blinking_on.cancel()
                        self.run_window.blinking_off.cancel()

                        self.run_window.blinking_on()
                        self.run_window.blinking_off()

                        timeout = time.time() + 5  # 5 seconds from now
                        while self.micro_sim.is_running:
                            try:
                                self.micro_sim.run_micro_instructions(timeout)
                                self.run_window.inst_table.get_data(self.micro_sim.index,
                                                                    self.micro_sim.disassembled_instruction())

                                if self.micro_sim.prev_index == self.micro_sim.index:
                                    self.micro_sim.is_running = False
                                else:
                                    self.micro_sim.prev_index = self.micro_sim.index
                            except (SystemError, TimeoutError) as e:
                                self.micro_sim.is_running = False
                                toast_message = f'Error! {e}'
                    self.run_window.reg_table.get_data()
                    self.run_window.mem_table.data_list.clear()
                    self.run_window.mem_table.get_data()
                    toast(toast_message)
```

`def assembler(self)` m is not being used. Change for loop to a while loop
```Python
            for m in range(50):
                f.write(f'{RAM_ASSEMBLER[i]} {RAM_ASSEMBLER[i + 1]}' + '\n')
                i += 2
```

ChristianPerez34 commented:
`def clear_dialog(self, instance)`, `def clear_decision(self, *args)`, `def clear(self)`, `def clear_run(self)`, `def open_editor_save_dialog(self, instance)`, `def save_asm_file(self, *args)`, `def buttons_information(self, instance)`, `def io_config_open(self, instance)`, `def save_io_ports(self, *args)`, `def file_manager_open(self, instance)`, `def get_data(self)` missing docstring

`def save_file` m is not used. Change for loop to a while loop
```Python
            for m in range(50):
                f.write(f'\n{RAM[i]}    {RAM[i + 1]}')
                i += 2
```

ChristianPerez34 committed:
Reformatted code. Optimized imports. added docstring for [HexKeyboard].
Added docstring for [write_ram] method

ChristianPerez34 committed:
Removed traffic light blinking events from [HexKeyboard]

cfboy committed:
Change for loop to whiles. Make change requests.

ChristianPerez34 committed:
Reformatted code and removed redundant parenthesis

ChristianPerez34 committed:
Merge branch 'NG_49_cleanup_optimization' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

cfboy committed:
Change gobal variables to upper case naming style.

cfboy committed:
Merge remote-tracking branch 'origin/NG_49_cleanup_optimization' into NG_49_cleanup_optimization

cfboy committed:
Delete unused variable.

cfboy committed:
Change color highlight on register table.

ChristianPerez34 committed:
Created directory to store IO device related functionality. Moved hex keyboard top IO module. Created events dictionary to ensure previous app functionality

ChristianPerez34 committed:
Merge remote-tracking branch 'origin/NG_49_cleanup_optimization' into NG_49_cleanup_optimization

Commit:
Moved TrafficLights to IO module.
Modified traffic light structure.
Modified IO events to be triggered on simulator run/debug

Commit:
Moved seven segment display to IO module.
Modified structure of seven segment display.

Commit:
Moved ascii grid to IO module

Commit:
Defined a single python file to hold all IO devices

ChristianPerez34 committed:
Merge branches 'NG_49_cleanup_optimization' and 'master' of https://github.com/needsguidance/Semref into NG_49_cleanup_optimization

Commit:
Added a TODO
